wgengine/magicsock/magicsock.go

```diff
@@ -1876,6 +1876,12 @@ func (c *Conn) Close() error {
 	}
 	defer c.mu.Unlock()
 
+	// activeDerp is non-nil iff at least one connection was opened.
+	// We must wait on derpStarted to ensure the derphttp client goroutine has exited.
```

Comment:
"the" derphttp client goroutine? There are up to 3. What are we actually waiting for?

But this Close method hasn't yet done anything to signal to those goroutines to start shutting down. Shouldn't we do the connCtxCancel() call first?

Reply:
I mean specifically the one spawned on line 1019. The `firstDerp` condition seems to be true at most once to me:
```go
1017	if firstDerp {
1018		startGate = c.derpStarted
1019		go func() {
1020			dc.Connect(ctx)
1021			close(c.derpStarted)
1022		}()
1023	}
```

Reply:
But you're right, it makes more sense to put it after `connCtxCancel`, otherwise we're waiting as long as it takes for it to connect.

Reply:
Okay, but provide some context for readers in the future. There's no description of what we're actually waiting for, no bug number, etc.

"We must wait on derpStarted to ensure the derphttp client goroutine has exited."  But why? What is that goroutine doing or holding or mutating?

Why just waiting on that one goroutine and not all of them? How does it know to stop? We're just going to wait 5 or 10 or 15 seconds for its timeout if it's blocked in a slow connect? We're not going to help it fail quickly by calling connCtxCancel?